### PR TITLE
[Feature] Ngon Radar

### PR DESCRIPTION
**Pull request by z64555**
This features adds the Ngon radar gauge, which is a derivative of the standard radar gauge. Instead of the standard's oval plot area, the Ngon radar gauge has a polygonal plot area.

The projection algorithm is the same as the standard's radar, so all the Ngon does is clamp the blip to the edges of the plot area.

I also refactored the standard radar plotting code a bit, there was some clamping in the original code that didn't work right, so blips could be drawn on top of the oval's edges. I'm not counting this as a bugfix, though.

Included is a test .tbm that activates a hexagonal plot area with "points up."

[radar_test-hdg.tbm.txt](https://github.com/scp-fs2open/fs2open.github.com/files/767768/radar_test-hdg.tbm.txt)

### CODE REVIEW

**Comment by z64555** [comment]
I put the unfinisehd tag on because I haven't tested the behavior of the radar icons yet, which probably needs a bit more refactoring of the stardard radar.

File: code/radar/radar.cpp [hunk 1]
```diff
@@ -367,43 +388,25 @@ void HudGaugeRadarStd::plotBlip(blip *b, int *x, int *y)
...
-		max_radius = i2fl(Radar_radius[0] - 5);
-
-		if (hypotenuse >= max_radius)
-		{
-			clipped_x_dist = max_radius * (new_x_dist / hypotenuse);
-			clipped_y_dist = max_radius * (new_y_dist / hypotenuse);
-			new_x_dist = clipped_x_dist;
-			new_y_dist = clipped_y_dist;
-		}
+		new_pos = ZERO_VECTOR;
```
Comment (The-E) [inline code/radar/radar.cpp]: For some unfathomable reason, this line breaks MSVC 2015. Apparently, you can only assign ZERO_VECTOR to something as part of its declaration (i.e. vec3d x = ZERO_VECTOR; is fine, the code above breaks things).

The easiest solution would be to initialize new_pos to ZERO_VECTOR and invert the zdist check.

File: code/math/vecmat.cpp [hunk 2]
```diff
@@ -1208,6 +1208,31 @@ float find_nearest_point_on_line(vec3d *nearest_point, const vec3d *p0, const ve
...
+	// Vector v2 forms an edge between v0 and v1, thus forming a triangle.
+	// An intersection exists between v0 and v1 if their cross product is parallel with the cross product of v2 and v1
+	// The scalar of v0 can then be found by the ratio between the two cross products
+	vec3d v2, crossA, crossB;
+
+	vm_vec_sub(&v2, p1, p0);
+	vm_vec_cross(&crossA, v0, v1);
+	vm_vec_cross(&crossB, &v2, v1);
+
+	if (vm_vec_equal(crossA, ZERO_VECTOR)) {
```
Comment (The-E) [inline code/math/vecmat.cpp]: Similarly, ZERO_VECTOR isn't going to work in this context.

File: code/radar/radar.cpp [hunk 3]
```diff
@@ -359,6 +359,27 @@ void HudGaugeRadarStd::pageIn()
 	bm_page_in_aabitmap( Radar_gauge.first_frame, Radar_gauge.num_frames );
 }
 
+void HudGaugeRadarStd::clampBlip(vec3d* blip)
```
Comment (The-E) [inline code/radar/radar.cpp]: "blip" is a struct defined in radarsetup, gcc and clang both think that choosing a different name for this parameter would be good (Or, alternatively, you could pass in the actual blip here?)

File: code/radar/radar.cpp [hunk 4]
```diff
@@ -359,6 +359,27 @@ void HudGaugeRadarStd::pageIn()
...
+{
+	float max_radius;
+	float hypotenuse;
+
+	max_radius = i2fl(Radar_radius[0] - 5);
+
+	// Scale blip to radar size
+	vm_vec_scale(blip, i2fl(Radar_radius[0])/2.0f);
+
+	hypotenuse = _hypotf(blip->xyz.x, blip->xyz.y);
```
Comment (The-E) [inline code/radar/radar.cpp]: "_hypotf" does not exist on *NIX, the canonical name is just "hypotf"

File: code/radar/radar.cpp (hunk 1 above)
Comment (The-E) [inline code/radar/radar.cpp]: It should be noted that VS2017 fixes this issue.

File: code/radar/radarngon.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
+#include "radar/radarngon.h"
+
+#include "math/vecmat.h"
+
+#include <math.h>
+
+void HudGaugeRadarNgon::clampBlip(vec3d* blip) {
```
Comment (The-E) [inline code/radar/radarngon.cpp]: This blip is still causing issues for the CI bots

File: code/radar/radarngon.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	float hypotenuse;
+
+	hypotenuse = _hypotf(blip->xyz.x, blip->xyz.y);
```
Comment (The-E) [inline code/radar/radarngon.cpp]: _hypotf must be changed to hypotf.

File: code/radar/radarngon.cpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	v->xyz.z = 0.0f;
+
+	vm_vec_sub2(v, p);
+}
+
+// Default to a hexagon
+HudGaugeRadarNgon::HudGaugeRadarNgon()
+	: arclen(PI2 / i2fl(6)), offset(0.0f), r_min(cos(PI / i2fl(6))) {}
+
+HudGaugeRadarNgon::HudGaugeRadarNgon(int num_sides, float offset)
```
Comment (The-E) [inline code/radar/radarngon.cpp]: "offset" shadows a member variable

**Comment by FSCyborg** [comment]
z64555 has let me finish up this PR.  Turns out the changes were already on the local branch, just not on the pull request.  The only actual work that I can claim is reversing some white space changes and rebasing it to the current code to make sure there weren't merge conflicts.  

I will test this soon (tm) to make sure it works.

**Comment by FSCyborg** [comment]
Apparently need to redo the cmake config before testing.  I will give this a try this weekend.  

Also need to double check that correct angle units are used, as per z64555's comment on discord.

**Comment by FSCyborg** [comment]
Having trouble compiling, so I haven't been able to test yet.  But I am adding changes that asarium suggested.

**Comment by FSCyborg** [comment]
Fixed the compilation issue (always remember to update your submodules). About to compile and test.  Let's see what happens!!

**Comment by Goober5000** [comment]
@FSCyborg what were the test results?

@The-E can you re-review based on FSCyborg's latest changes?

**Comment by FSCyborg** [comment]
So far tests are showing no gauge differences.

**Pull request merged**